[1단계 방탈출 결제 / 배포] 밍곰(박민선) 미션 제출합니다.

minSsan opened the pull request:
안녕하세요 찰리 ~! 이번 리뷰이로 배정받은 밍곰이라고 합니다! 🙂

코드 변경 범위는 다음과 같습니다!
- [변경사항](https://github.com/woowacourse/spring-roomescape-payment/pull/211/files/ad382f786bcf4ec5747290a2e2d8cca6dc6af30a..e59360d91037c47cf26520bc7b1ea23fd9cffe6f)

질문은 코멘트로 남기겠습니다! 잘 부탁드립니다 🙇🏻‍♀️

---

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  [방탈출 결제](https://prolog.techcourse.co.kr/studylogs/4476)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

minSsan committed:
setup: 예약 대기 구현 import

minSsan committed:
setup: 클라이언트 코드에 결제 요구사항 추가

minSsan committed:
feat: 클라이언트 토스 API prefix 설정

minSsan committed:
feat: 토스 페이먼츠 API 연동 클래스 구현

minSsan committed:
feat: Payment 엔티티 정의

minSsan committed:
feat: PaymentRepository 정의

minSsan committed:
feat(CreateReservationService): createWithPayment 메서드 구현 - 멤버 예약 생성시 결제 요청

minSsan committed:
feat: ReservationPaymentClient 객체 빈 등록을 위한 Configuration 파일 정의

minSsan committed:
test(CreateReservationService): 외부 API 객체(ReservationPaymentClient) 주입 및 예약 생성 테스트를 어드민 권한으로 변경

minSsan committed:
feat(ReservationController): 멤버의 예약 생성 메서드 변경 - 결제 예약

minSsan committed:
test(ReservationController): 외부 API 의존성 추가로 인한 Mocking 추가

minSsan committed:
test(CreateReservationService): 멤버 결제 예약 생성 테스트 작성

- 외부 의존성 Mocking 하여 단위 테스트로 진행

minSsan committed:
refactor: time, theme 테스트에 필요한 Reservation 생성 요청을 어드민 요청으로 변경

minSsan committed:
docs(README.md): 요구사항 문서 추가

minSsan committed:
refactor: 토스 페이먼츠 결제 서비스 네이밍 변경 및 base url 설정 파일 추가

minSsan committed:
chore: 미사용 테스트 삭제

minSsan committed:
refactor(CreateReservationService): 검증 로직 수정 및 예약 생성시 `@Transactional` 추가

minSsan committed:
feat(ReservationService): delete 메서드에 `@Transactional` 추가 및 WaitingRepository 메서드 시그니처 수정

minSsan committed:
chore: TODO 주석 추가

minSsan committed:
docs(README.md): 계정 정보 수정

minSsan committed:
refactor: 결제시 payment 생성 책임을 TossPaymentService 내부로

minSsan commented on src/main/java/roomescape/reservation/service/CreateReservationService.java:
```diff
@@ -77,4 +91,14 @@ private void validateReservationDateTime(Reservation reservation) {
             throw new IllegalArgumentException("과거 날짜의 예약은 생성할 수 없습니다.");
         }
     }
+
+    @Transactional
+    public ReservationResponse createWithPayment(ReservationWithPaymentRequest request, LoginMember loginMember) {
+        ReservationCreateRequest reservationCreateRequest = ReservationCreateRequest.from(request, loginMember);
+
+        ReservationResponse reservationResponse = create(reservationCreateRequest);
+        tossPaymentService.postConfirmPayment(ConfirmPaymentRequest.from(request));
+
+        return reservationResponse;
+    }
```
현재 코드에서는 결제 서비스(`TossPaymentService`)를 `CreateReservationService`가 직접 호출하고 있어요
> 개인적으로 **서비스가 서비스를 참조하는 방식**을 선호하지 않는 편인데요 ,, 🥹 일단 리팩터링 없이 PR을 남기게 되었어요

### 결제와 예약 생성을 하나의 트랜잭션으로 관리
> 일단 이렇게 구현하게 된 이유는, `결제`와 `예약`을 하나의 트랜잭션으로 관리하려는 목적이에요 ,,!

**(1) 결제를 성공**하고, 이에 따라 **(2) `Reservation` 엔티티를 저장**하는 과정은 하나의 트랜잭션에서 관리되어야 한다고 생각했습니다.
그래서 해당 메서드에 `@Transactional` 어노테이션을 붙였는데요, 

문득 _"**외부 API**를 호출하는 부분까지 **트랜잭션에 포함**되는 것이 괜찮을까?"_ 하는 생각이 들었어요 🤔

- 외부 API 호출은 **시간 지연**이 발생하는 작업
    - 만약 이 부분까지 트랜잭션으로 묶어버리게 되면 `예약 생성` 요청마다 **트랜잭션에 필요한 자원을 오래 붙들게 되지 않을까** 라는 생각이 들었어요
- 외부 API 호출 과정에서는 **DB 작업**이 존재하지 않음
    - 사실상 `결제 API를 호출하고 응답을 받는 과정`에서는 DB 작업이 불필요해요.

### `TossPaymentService`에서 DB 작업을 분리하기 ...?
지금 당장에는 해당 서비스 객체에서 **`PaymentRepository` 의존성을 제거**하고, **결제 API를 호출하고 응답을 받는 역할만** 하도록 하는 방식이 떠오르는데요.
> 이 경우에는 `service` 계층보다는 `infrastructure` 라는 별도의 패키지로 분리할 것 같아요.

하지만 이 경우에도 아래 코드처럼 `컨트롤러에서 따로 호출`한다고 하면, 두 과정을 분리할 수는 있겠지만 트랜잭션이 하나로 묶여있지 않기 때문에 본래 트랜잭션을 사용했던 목표를 이룰 수 없을 것 같습니다 ㅠ

```java
// ReservationController.java

@PostMapping
public ResponseEntity<ReservationResponse> createWithPayment(
        @Valid @RequestBody ReservationWithPaymentRequest request,
        final LoginMember loginMember
) {
    // 💥 1번 ~ 3번 작업은 같은 트랜잭션에서 관리되지 않기 때문에, 작업의 원자성이 보장되지 않음

    // 1. 요구사항에 맞는 Reservation이 성공적으로 생성되는지 확인
    ReservationResponse reservationResponse = createReservationService.createWithPayment(request, loginMember);

    // 2. 결제 승인 API 호출
    tossPaymentClientApi.postConfirmPayment(ConfirmPaymentRequest.from(request));

    // 3. 결제 승인 응답을 기반으로 payment 저장
    // ...
    return ResponseEntity.ok(reservationResponse);
}
```

---

`외부 API`를 **트랜잭션 범위에서 분리**하고 싶은데, 결제 승인 / `Reservation` 생성 과정 / `Payment` 생성 과정이 함께 이뤄져야 해서 애초에 분리가 가능한건지 모르겠습니다 ..

찰리는 이에 대해서 어떻게 생각하시나요?!
`외부 API 호출을 트랜잭션에서 분리`할 필요성이 있다고 생각하시나요?!

Gomding replied:
트랜잭션에서 외부 요청은 분리하는것이 좋다고 생각해요!
말씀해주신 이유가 크리티컬 하다고 생각하구요

예약 커밋 후 -> 결제 진행 이렇게 분리할 수 있는데
핵심은 결제 실패했을 때 어떻게 처리해야할까 같네요 😃 
해당 부분을 집중해서 생각해보죠 ㅎㅎ
꼭 트랜잭션 롤백만이 정답일까요?

minSsan replied:
🤔🤔🤔🤔🤔🤔 
이 부분에 대해서 고민을 많이 하다보니 2차 리뷰 요청이 많이 늦어졌아요 .. 죄송합니다 😓 
> _(사실 지금 제출한 코드도 잘 설계된 방향인지 모르겠습니다 ..😢)_

---

찰리가 말씀해주신대로, 모든 과정을 트랜잭션으로 묶는 대신, **`DB 작업을 포함한 로직`을 통해서도 충분히 API 요청이 실패했을 경우에 대처가 가능**할 것 같아요!

하지만, 결제 승인 API 요청을 처리하는 과정에서 오류가 발생하는 경우 `어떤 상태에서 오류가 발생했는지` 나타내는게 중요할 것 같다는 생각이 들었어요.
그래서 `Payment`에 아래와 같은 상태 값(`PENDING`, `COMPLETE`, `FAILED`)을 추가했습니다.
> **결제에 실패한 경우에는** 환불 절차가 필요하기 때문에, `Payment` 객체를 **삭제하지 않고** 상태 값(`FAILED`)으로 이를 나타내는 것이 좋겠다고 생각했어요 🤔

```java
// PaymetStatus.java
public enum PaymentStatus {
    PENDING(),
    COMPLETE(),
    FAILED(),
    ;
}

// Payment.java
@Entity
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String paymentKey;

    private Integer amount;

    @Enumerated(value = EnumType.STRING)
    private PaymentStatus status;

    private Long reservationId; // 결제가 실패하는 경우에는 Reservation이 존재하지 않으므로, id 값으로만 연관관계를 설정함

    // ...
}
```


### 예약 및 결제 승인 흐름
1. `Reservation` + `Payment` (`PENDING`) 생성 후 커밋 _(with 트랜잭션)_
2. 결제 승인 API 호출 _(**without 트랜잭션**)_ 👉🏻 DB 작업 분리
3. `ReservationService`에서 결제 승인 API 응답에 따라 작업 처리 _(with 트랜잭션)_
    a. 결제 실패 - 생성된 `Reservation` 삭제 & `Payment`의 상태를 **실패 상태(`FAILED`)** 로 표시
    b. 결제 성공 - `Payment`의 상태를 **완료 상태(`COMPLETED`)** 로 표시

#### 적용 코드
```java
// CreateReservationWithPaymentService.java

// 트랜잭션 X
public ReservationWithPaymentResponse create(ReservationWithPaymentRequest request, LoginMember loginMember) {
    // 1. Reservation 엔티티 + Pending 상태의 Payment 엔티티 생성
    // 👉🏻 CreateReservationService.createWithPendingPayment가 트랜잭션 내에서 실행된다.
    ReservationWithPaymentResponse reservationWithPaymentResponse = createReservationService.createWithPendingPayment(request, loginMember);

    // 2. 토스 페이먼츠 API 호출
    // 👉🏻 트랜잭션 X
    ResponseEntity<ConfirmPaymentResponse> confirmPaymentResponse = tossPaymentClient.postConfirmPayment(ConfirmPaymentRequest.from(request));

    // 3-1. 결제 승인이 성공한 경우, Payment 엔티티의 상태를 Completed로 변경
    if (confirmPaymentResponse.getStatusCode().is2xxSuccessful()) {
        paymentService.completePayment(reservationWithPaymentResponse.paymentId());
        return reservationWithPaymentResponse;
    }

    // 3-2. 결제 승인이 실패한 경우, Reservation 엔티티를 삭제하고 Payment 엔티티의 상태를 Failed로 변경
    deleteReservationService.delete(reservationWithPaymentResponse.id(), loginMember);
    paymentService.failedPayment(reservationWithPaymentResponse.paymentId());
    
    // 3-2-1. 예외처리
    tossPaymentClient.validateResponse(confirmPaymentResponse);
    throw new InternalServerErrorException();
}
```

---

### 관련 커밋 내역

#### 토스 페이먼츠 API 전송 객체에서 DB 작업 분리
[feat: TossPaymentService -> TossPaymentClient 변경 및 @Retryable 적용](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/117e90e0d07a0bbe80ae94435a28673289fee83e)

#### 외부 API 호출을 포함한 결제 예약을 생성하는 서비스 클래스 정의
[feat: 결제 예약 생성 요청을 담당하는 CreateReservationWithPaymentService 객체 정의](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/409b59f9b2f0c55be02e453e20c88d04edeb5946)

#### 서비스 계층 분리
[feat: ReservationService -> DeleteReservationService, ReservationQuer…](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/a47ef6ae188d424e16e3782d266eaa5caf6aeb2c)

> 웬만하면 서비스 계층을 나눌 생각이 없었는데요, 😳
> 
> 결제+예약의 전체 과정을 **컨트롤러가 아닌 서비스 객체 내에서 처리**하려고 하다보니, 트랜잭션 분리를 위해서는 서비스 계층이 필요하다는 판단이 들었어요.   
> - 스프링 트랜잭션은 프록시 **'객체'** 를 만들어서 처리하기 때문에, 원하는 작업 _(`Reservation`, `Payment` 생성 / 삭제 / 수정)_ 에만 선택적으로 트랜잭션을 걸기 위해서는 **별도의 클래스를 정의해서 호출하는 방식**이 가장 간단하더라구요 ..!

---

아래 커밋의 테스트 코드를 보시면 `CreateReservationWithPaymentService` 가 **토스 API 응답에 따라서 어떤식으로 예약(`Reservation`) 및 결제(`Payment`)를 처리하는지** 조금 더 쉽게 이해하실 수 있을 거예요..!

[test(CreateReservationWithPaymentService): 예약 생성 및 결제 내역 생성 테스트 작성](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/257b812e390bbcb403ec7b37034d6cb6e1b425b0)

Gomding replied:
허허 이정도로 딥다이브 할 줄은 몰랐는데 고생하셨습니다 😮 
치열하게 고민하신게 느껴지네요 ㅎㅎ

현재 상황에서 적절하게 구현해주셨다고 생각해요!

minSsan replied:
🫨🫨🫨🥹🥹🥹🥹🥺🥺🥺🥺
감사합니다 🙇🏻‍♀️!!

Gomding commented on README.md:
```diff
@@ -0,0 +1,86 @@
+| `/themes`       | GET | 테마 조회    |
+| `/times`        | GET | 예약 시간 조회 |
+
+### 토스 페이먼츠 API 결제 예외 처리
+
+아래 두 예외를 제외한 나머지 예외는 토스 API 예외를 그대로 반환한다.
+
+- `INCORRECT_BASIC_AUTH_FORMAT`
+- `INVALID_API_KEY`
```
어떤 기준으로 정해진 것인지 궁금해요! 👍

minSsan replied:
넵 일단 분류 기준은 [토스 페이먼츠 API - 결제 승인 에러 코드](https://docs.tosspayments.com/reference/error-codes#%EA%B2%B0%EC%A0%9C-%EC%8A%B9%EC%9D%B8) 문서를 확인하면서 결정했어요~

토스 API 에서는 저희 어플리케이션을 `클라이언트`로 바라보기 때문에, **어플리케이션에서 토스 API 요청 코드가 잘못 작성된 경우**에도 `400` 에러를 반환하는 경우가 몇 가지 있었어요!

하지만 이 경우에는 **저희 어플리케이션의 클라이언트가 입력을 잘못 보낸 것이 아닌**, **방탈출 서버 측의 코드가 잘못된 것**이 원인이기 때문에, 클라이언트에게는 `400` 에러보다는 `5XX` 에러를 반환하는 것이 적절하다고 생각했습니다!

그리고 리드미에 명시한 코드들이 위의 경우에 해당한다고 판단해서 `Internal Server Error (500)` 으로 처리했습니다!

- `INVALID_API_KEY`   
**서버 측에서 시크릿 키를 잘못 입력한 경우**에 반환되는 값이므로, 클라이언트가 입력 값을 잘못 입력했다기보다는 **서버 내부 코드가 잘못된 경우**라고 생각했습니다
- `INCORRECT_BASIC_AUTH_FORMAT`   
서버 측에서 **Authorization 헤더의 인코딩을 잘못한 경우**에 발생하므로, 마찬가지로 서버 내부 에러로 표현했습니다.
- `INVALID_AUTHORIZE_AUTH`   
**서버 측에서 토스 API에 명시된 인증 방식을 따르지 않은 경우**에 발생하므로 서버 내부 에러로 표현했습니다.

Gomding replied:
적절한 기준으로 잘 처리해주셨네요 💯 💯 💯

Gomding commented on README.md:
```diff
@@ -0,0 +1,86 @@
+# 사전 생성 계정 정보
+- 관리자 정보
+    - 이메일: minggom@woowa.com
+    - 비밀번호: 12341234
+
+- 사용자 정보
+    - 이메일: posty@woowa.com
+    - 비밀번호: 12341234
+
+# 🚪방탈출 애플리케이션
+
+`RoomescapeApplication`에서 실행 가능합니다.
```
실행 정보 명시 👍 👍 👍 👍 
README 를 잘 활용해주셨네요!!

minSsan replied:
감사합니다! 🙂

Gomding commented on src/main/java/roomescape/payment/domain/Payment.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String paymentKey;
+
+    private Integer amount;
```
다른 값은 필요없을지 고민해볼 수 있겠네요 :)

minSsan replied:
결제 정보와 관련해서는 깊게 고민해보지 않았는데요 🤔

이번 [피드백 과정](https://github.com/woowacourse/spring-roomescape-payment/pull/211#discussion_r2120935543)에서 `결제 상태` 및 `예약에 대한 참조`가 필요할 것이라는 생각이 들어서 아래와 같이 두 개의 필드를 추가해줬어요~!

https://github.com/woowacourse/spring-roomescape-payment/blob/d77c095f8ff4cfae3d3e328930abacba47ec7ce4/src/main/java/roomescape/payment/domain/Payment.java#L22-L25

Gomding commented on src/main/java/roomescape/payment/service/TossPaymentService.java:
```diff
@@ -0,0 +1,66 @@
+    @Value("${api.toss.secret-key}")
+    private String secretKey;
+
+    private final RestClient restClient;
```
다른 HTTP Client 도 있는데
RestClient 를 선택하신 이유가 궁금해요~

어떤 것을 고려했고 장단점 비교한 부분을 공유부탁드려요 😃

minSsan replied:
우선 페어 프로그래밍이 진행된 시점에서는 학습 테스트를 통해 `RestClient`, `RestTemplate` 만을 사용해본 상태였어요. 🥹
학습 테스트 과정을 통해, 둘 중에서는 `RestClient`가 더 **직관적**이고 **가독성이 높은** 것 같아서 사용했습니다!

### `WebClient`
> 를 적용하고 싶었지만, 러닝 커브에 비해 제가 가진 시간이 부족해서 ... 적용하진 못했습니다 😓

리뷰 피드백 과정에서 **`WebClient`와 `RestClient`&`RestTemplate` 의 차이점**에 대해서도 간단하게 조사해봤는데요.

`RestClient`&`RestTemplate`와 다르게, `WebClient`는 **논블로킹 방식**으로 동작하기 때문에 현재 상황에서는 **`WebClient`를 사용하는 것이 더 적합할 수 있겠다**는 생각이 들었어요.

> 블로킹의 경우에는 응답을 받을 때까지 스레드를 반환하지 않고, 스레드가 아무런 일을 하고있지 않은 상태 그대로 대기하고 있기 때문이에요.

논블로킹은 요청을 보내자마자 일단 스레드를 반환하고, 응답을 받으면 스레드를 다시 할당받아서 요청을 처리하기 때문에 **스레드 자원을 조금 더 효율적으로 쓸 수 있을 것 같다**는 생각이 들었어요.

- 만약 외부 API의 응답 시간이 길어진다면, 블로킹의 경우 **스레드 자원을 불필요하게 점유하는 시간이 길어질 수 있을 것 같다**는 생각이 들었습니다 ...!

---

[공식 문서](https://docs.spring.io/spring-framework/reference/integration/rest-clients.html) 상에서는 
- `RestClient`와 `RestTemplate`는 **synchronous** 하다
- `WebClient`는 **non-blocking**이다

정도로만 설명을 하고 있어서, 사실 **둘의 차이를 블로킹/논블로킹으로 이해해도 되는건지** 잘 모르겠습니다 .. 학습하면서도 많이 헷갈렸는데, 사실 지금도 정확히 이해하진 못한 것 같아요 ㅜ

Gomding replied:
스프링 HTTP Client 의 흐름을 파악해보셔도 좋을 것 같네요 :)

- https://www.youtube.com/watch?v=S4W3cJOuLrU

Gomding commented on src/main/java/roomescape/payment/service/TossPaymentService.java:
```diff
@@ -0,0 +1,66 @@
+@Service
+public class TossPaymentService {
+    private static final List<String> IGNORE_CODES = List.of(
+            "INCORRECT_BASIC_AUTH_FORMAT",
+            "INVALID_API_KEY",
+            "INVALID_AUTHORIZE_AUTH"
```
README 내용과 차이가 있군요 😎 

```
아래 두 예외를 제외한 나머지 예외는 토스 API 예외를 그대로 반환한다.

- `INCORRECT_BASIC_AUTH_FORMAT`
- `INVALID_API_KEY`
```

minSsan replied:
허허 문서에도 같이 반영한다는걸 깜빡했네유 .. 😓

[docs(README.md): 예외 핸들링 제외 케이스 추가](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/f1810ecac81ce13763356d47274b434d5d747d7d)

Gomding commented on src/main/java/roomescape/payment/service/TossPaymentService.java:
```diff
@@ -0,0 +1,66 @@
+            return ;
+        }
+        PaymentFailure failure = response.failure();
+        if (IGNORE_CODES.contains(failure.code())) {
+            throw new RuntimeException(failure.message());
+        }
+        throw new PaymentClientException(failure.message());
```
CODE 에 따라 어떤 차이가 있는걸까요?

minSsan replied:
우옷 ... 질문이 따로 있었네요..!
다른 [코멘트](https://github.com/woowacourse/spring-roomescape-payment/pull/211#discussion_r2117793243)에 관련해서 설명 적어두었습니다

Gomding commented on src/main/java/roomescape/payment/service/TossPaymentService.java:
```diff
@@ -0,0 +1,66 @@
+            @Value("${api.toss.url}") String baseUrl,
+            PaymentRepository paymentRepository
+    ) {
+        this.restClient = RestClient.builder().baseUrl(baseUrl).build();
```
RestClient 에 baseUrl 외에 다른 설정은 필요없을까요?
어떤 설정들이 있는지 살펴보면 좋겠어요~

minSsan replied:
`RestClient`를 빈으로 등록하는 과정에서, **`baseUrl`을 포함한 `Auth 헤더` 설정, `read timeout` 값** 설정을 추가했어요..!
> 외부 API인 만큼 요청 처리 시간 초과에 대해서도 설정 값을 넣어주는 것이 좋을 것 같아용

`connection timeout` 값도 넣어주고 싶었는데, 어느 정도의 시간 값으로 설정해야할지 잘 모르겠어서 일단 보류했습니다 .. 😓

---

### 관련 커밋
#### RestClientConfig 설정 및 RestClient 설정 추가
https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/56bb459bf25709c2290af283a9da101b84b00107#diff-c2cbc415a8e571ee06c1a5f6f5dfa601aebc1b8bd06073ee22a7592680c2b983

커밋 내역에는 10초로 설정되어 있는데, [관련 토스 API 공식문서](https://docs.tosspayments.com/resources/glossary/timeout#read-timeout)를 토대로 `15초`로 변경했어요 ..!
> [`TossPaymentClient`](https://github.com/woowacourse/spring-roomescape-payment/blob/51b22ba338cf22a88aeca2604a9e8ca6a02577e4/src/main/java/roomescape/infrastructure/TossPaymentClient.java)에서 요청 Retry를 3번(default) 해주고 있고, 15초씩 3번 요청하게 되면 총 45초 (30초 ~ 60초) 

#### 멱등키 설정

[`TossPaymentClient` - 멱등키 헤더 설정](https://github.com/woowacourse/spring-roomescape-payment/blob/d77c095f8ff4cfae3d3e328930abacba47ec7ce4/src/main/java/roomescape/infrastructure/TossPaymentClient.java#L32-L35)

요청 재시도(`@Retryable` 적용)로 인해서 중복 요청이 발생해요. 
중복해서 결제가 진행되지 않도록 [토스 API 공식 문서](https://docs.tosspayments.com/reference/using-api/authorization#%EB%A9%B1%EB%93%B1%ED%82%A4-%ED%97%A4%EB%8D%94)를 참고해서 멱등키 헤더를 추가해봤어요

Gomding replied:
> connection timeout 값도 넣어주고 싶었는데, 어느 정도의 시간 값으로 설정해야할지 잘 모르겠어서 일단 보류했습니다 .. 😓

해당 설정값도 적절한 값으로 설정해보시면 좋겠어요 :)

> 커밋 내역에는 10초로 설정되어 있는데, [관련 토스 API 공식문서](https://docs.tosspayments.com/reference/using-api/authorization#%EB%A9%B1%EB%93%B1%ED%82%A4-%ED%97%A4%EB%8D%94)를 토대로 15초로 변경했어요 ..!

링크주신 곳에서 타임아웃에 대한 내용은 찾지 못했어요 😢

minSsan replied:
아앗..... 링크를 잘못 첨부했네요 😓
원래 첨부하려던 링크는 아래 링크가 맞습니다..!

[토스 API - 타임아웃](https://docs.tosspayments.com/resources/glossary/timeout#read-timeout)

minSsan replied:
### Connection Timeout

TCP 연결을 위해 **3 way handshake** 과정을 거치게 되는데, 이 과정이 실패하는 경우를 대비해 적절한 `timeout` 값을 설정해야 해요.

<img width="443" alt="image" src="https://github.com/user-attachments/assets/8b5189c0-95d8-475f-9058-9d09311fe9b3" />

> 기본적으로 `timeout` 값이 설정되어 있지 않기 때문에, 만약 **3 way handshake**가 실패하는 경우, 서버는 응답을 무한정 기다리게 될 것이고 이는 유저에게 서버가 멈춘 것처럼 보일 것 같아요. 
> 따라서 `connection timeout` 값을 적절히 설정해주는 것도 매우 중요하다는 것을 느낄 수 있었습니다!

### 값 설정하기

다른 크루들과 이야기도 해보고, 관련 자료도 몇 가지 검색해보니 `RTT` _(= 하나의 요청을 보내고 응답을 받기까지 걸리는 시간)_ 를 토대로 `connection timeout` 값을 설정하는 것 같아요.


#### RTT 측정
사실 서버를 아직 배포한 상태가 아니라서, 현재 상황에서는 직접 측정하는 것이 무의미한 것 같긴 합니다만.. 😓
그래도 일단 `ping`으로 대략적인 요청~응답 시간을 측정해봤어요..!

```
--- https://api.tosspayments.com ping statistics ---
23 packets transmitted, 23 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 7.267/10.015/11.365/0.960 ms
```
> 이 과정에서 요청이 TCP가 아닌 다른 프로토콜 _(계층도 다름)_ 로 전달되기 때문에 정확하진 않지만, 평균 `10ms` 정도의 시간이 측정됐어요.

TCP가 더 상위 계층에 있음을 고려해서 RTT 값을 여유롭게 대략 `100ms`으로 잡고, **3 way handshake**를 고려해서 총 `300ms`으로 `connection timeout`을 설정할 수 있을 것 같아요.

여기에 시간을 더 보수적으로 잡아서 3의 배수로 측정하면 `900ms`, `약 1초`이기 때문에 최종적으로 **connection timeout**을 `1000ms`으로 설정했습니다..! _(1초가 너무 짧은 것 같기도 한데 🤔)_

---

[feat: timeout 설정을 yml 파일에서 관리하도록 변경](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/0ff895f692108d86a92c4b7065dfca4f1bfcfcad)
> **서버의 물리적인 환경에 따라서도 `timeout` 값이 달라질 수 있다**는 점을 알게 되어서, 이를 코드가 아닌 `yml` 설정 파일에서 관리하도록 변경했어요

[feat: toss api connection timeout 변경 - 3초 -> 1초](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/5e8bad6da5dcb4a797d8ba7bc45a5f308e3bfd75)

Gomding commented on src/main/java/roomescape/payment/service/TossPaymentService.java:
```diff
@@ -0,0 +1,66 @@
+            @Value("${api.toss.url}") String baseUrl,
+            PaymentRepository paymentRepository
+    ) {
+        this.restClient = RestClient.builder().baseUrl(baseUrl).build();
```
RestClient 를 빈으로 관리하는것은 어떻게 생각하세요?

minSsan replied:
사실 미션 중에는 이 부분에 대해서 크게 고민하지 못 했는데요, 🙄

찰리의 코멘트를 보고 생각해보니, 현재 코드에서는 `baseUrl`을 직접 주입받고 있어서 **`TossPaymentService` 객체의 단위 테스트가 매우 어려워질 것** 같아요 🤔
> `RestClient`에 대한 의존성이 내부에 숨겨져 있고, 따라서 외부(= 테스트 코드)에서는 이를 알 수 없기 때문

또한 **매 요청마다 필요한 값 _(ex. `Authorization` 헤더)_ 을 한 곳에서 관리**하는게 좋을 것 같습니다!
- 추후 토스 페이먼츠에 다른 요청이 필요한 경우에도 동일한 `secretKey` 값을 사용할텐데, 이를 매번 설정해주면 중복 코드가 늘어날 것 같아요.

찰리가 제안해주신 것처럼 **토스 API에 사용되는 `RestClient`를 빈으로 등록**하면, **(1) 단위 테스트 용이성**과 **(2) 중복되는 로직 방지**가 가능할 것 같아요~!

---

✚ 그리고 역할/책임 측면에서도 현재 `TossPaymentService` 객체에서 맡은 책임이 너무 과한 것 같기도 하네요 .. 🥹

[트랜잭션 문제](https://github.com/woowacourse/spring-roomescape-payment/pull/211#discussion_r2113565469)도 그렇고, 고민을 더 해보면서 기존에 `TossPaymentService` 객체가 맡았던 책임들을 아래와 같이 분산시키는게 좋을 것 같다는 생각이 들었어요 😳
|기존에 맡았던 책임|개선 방향|
|----|----|
|`RestClient` 요청 값 설정|`Configuration` 객체로 분리 + 의존성 주입 받기|
|토스 페이먼츠 결제 승인 API 호출|-|
|결제 승인 API 응답을 토대로 `Payment` 생성 및 저장|트랜잭션으로 관리되는 별도의 서비스 객체 메서드로 분리하기|

[feat: 토스 페이먼츠 API RestClient를 빈으로 등록](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/56bb459bf25709c2290af283a9da101b84b00107)

Gomding replied:
책임을 잘 분리해주셨네요 👍 👍 👍 
테스트 관점에서도 잘 고려해주셨구요 ㅎㅎ

밍곰의 의견을 자세하게 공유해주셔서 감사해요~ 💯 💯 💯

Gomding reviewed (changes requested):
안녕하세요 밍곰!
미션 잘 구현하셨네요 👍 
몇가지 코멘트 남겼으니 확인부탁드려요~

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요 🙏

minSsan committed:
feat: 예약 내역 필터링 검색 시 시작 날짜, 종료 날짜 검증 로직 추가

minSsan committed:
test(ReservationService): 예약 내역 필터링 검색 시 시작 날짜, 종료 날짜 검증 테스트 추가

minSsan committed:
feat(ReservationTimeService): create, delete 메서드 예외 수정

minSsan committed:
feat: 트랜잭션 관리가 필요한 메서드에 대해서 `@Transactional` 어노테이션 추가

minSsan committed:
refactor(MemberCreateRequest): dto -> entity 변환 메서드 추가

minSsan commented:
@coderabbitai pause

minSsan committed:
refactor(*): 커스텀 예외 수정 - ErrorCode enum 정의하여 일반화

minSsan committed:
refactor: 예외 응답 객체 추가

minSsan committed:
docs(README.md): 예외 핸들링 제외 케이스 추가

minSsan committed:
refactor(TossPaymentService): 토스 API 서버 에러 응답 메시지 삭제

minSsan committed:
feat: 토스 페이먼츠 API RestClient를 빈으로 등록

minSsan committed:
setup(build.gradle): spring retry 의존성 추가

minSsan committed:
feat: TossPaymentService -> TossPaymentClient 변경 및 `@Retryable` 적용

- TossPaymentClient: DB 의존성 삭제
- Toss API 호출시 멱등키 지정

minSsan committed:
feat: 결제 예약을 생성 요청을 담당하는 CreateReservationWithPaymentService 객체 정의

minSsan committed:
test(CreateReservationWithPaymentService): 예약 생성 및 결제 내역 생성 테스트 작성

minSsan committed:
fix: ReservationResponse 정적 생성자 메서드 누락

minSsan committed:
feat: ReservationService -> DeleteReservationService, ReservationQueryService로 분리

minSsan committed:
feat(RestClientConfig): 토스 페이먼츠 API read timeout 시간 변경 - 15초

minSsan committed:
refactor(CreateReservationWithPaymentService): 변수명 변경

minSsan committed:
feat(TossPaymentClient): `@Retryable` retryFor 삭제

minSsan committed:
fix(TossPaymentClient): 요청 재시도마다 UUID 재생성 - 멱등키를 파라미터로 주입받기

minSsan reviewed (commented):
안녕하세요 찰리~!
좋은 피드백 남겨주셔서 감사드립니다 🙇🏻‍♀️

빠르게 피드백 남겨주셨는데, 늦게 리뷰 요청 드려서 죄송합니다 ... 
고민하다보니 시간이 많이 소요돼서 리뷰 요청이 예상보다 늦어졌어요 🤯..ㅜ

리팩터링을 조금 크게(?) 해버려서 코멘트도 길어진게 있는데,, 미리 사과드립니다 😔
이번 피드백도 잘 부탁드려요 🙇🏻‍♀️💦

Gomding commented on src/main/java/roomescape/payment/domain/vo/PaymentStatus.java:
```diff
@@ -0,0 +1,8 @@
+public enum PaymentStatus {
+    PENDING(),
+    COMPLETE(),
+    FAILED(),
```
간결하게 표현할 수도 있겠군요! 😃 

```suggestion
    PENDING,
    COMPLETE,
    FAILED,
```

Gomding commented on src/main/java/roomescape/infrastructure/TossPaymentClient.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.UUID;
+
+@Component
+public class TossPaymentClient {
```
`TossPaymentClient` 해당 빈도 payment 패키지 내부에 있어야하지 않을까요!

minSsan replied:
옙! 🫡
처음에는 `infrastructure`를 `common`과 비슷한 개념으로 상위 패키지로 정의했었는데요,

외부 환경에 의존하는 객체가 많아지면 `infrastructure` 내부 구조도 복잡해질 것이고, 그만큼 도메인 별로 구별하기가 점점 더 어려워질 것 같네요,, 😥

해당 패키지를 `payment` 패키지 하위로 이동시켰습니다!

[refactor: TossPaymentClient 를 관리하는 infrastructure 위치 변경 - payment 도메인…](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/5b55c3965e5f8890309ff94b93abb58a6d0ad92a)

Gomding commented on src/main/java/roomescape/config/RestClientConfig.java:
```diff
@@ -0,0 +1,30 @@
+package roomescape.config;
```
다른 외부 API 연동도 필요할 텐데
해당 설정은 결제를 위한 RestClient 설정이 주를 이루고 있어요 :)
payment 패키지 내부에 위치하는것은 어떻게 생각하시나요?

minSsan replied:
일단 패키지는 **도메인별로 나눈** 것인데요!
`RestClientConfig`는 **외부 API에 요청을 보내는데 필요한 설정**을 하고 있어요.

이는 도메인 로직과는 거리가 조금 멀다고 생각해서, `payment` 패키지가 아닌 `roomescape.config` 패키지에 뒀어요 🤔
혹시 이에 대해서는 어떻게 생각하시나요?!

Gomding replied:
공통으로 관리한다면 상관 없다고 생각해요 :)

만약 해당 설정이 결제를 위한 Config 로 분리된다면 payment 패키지가 맞을 것 같구요 😃 
결국 `결제라는 도메인` 에서 사용되는 설정이니까요 ㅎㅎ
가장 관련있는 곳에 위치한다면 프로젝트에 대한 지식 없이도 예측하기 좋다고 생각해요!

Gomding commented on src/main/java/roomescape/config/RestClientConfig.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.web.client.RestClient;
+
+@Configuration
+public class RestClientConfig {
```
의도하신 부분인지 모르겠지만
`RestClientConfig` 는 범용적인 이름일 것 같은데
어떻게 생각하시나요!

앞으로 다른 외부 API 연동이 필요할 때 
다른 RestClient 빈을 생성하는 코드도 여기에 구현될 예정일까요?

minSsan replied:
아 넵! 
`Payment` 관련 API만 해당 객체에서 관리하려고 했는데, 객체 이름에 이 의도가 잘 드러나지 않은 것 같아요 😥

객체명 수정했습니다..!

[refactor: RestClientConfig -> PaymentRestClientConfig 이름 변경](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/157e435721dc0085bface0b8e2ed9f6754a7f402)

Gomding commented on src/main/java/roomescape/infrastructure/TossPaymentClient.java:
```diff
@@ -0,0 +1,72 @@
+                .toEntity(ConfirmPaymentResponse.class);
+    }
+
+    public void validateResponse(ResponseEntity<ConfirmPaymentResponse> response) {
```
`org.springframework.web.client.ResponseErrorHandler` 에 대해서도 학습해보시면 도움 될거에요 :)

minSsan replied:
생각해보니 `ResponseErrorHandler` 내부에서 **토스 페이먼츠 API의 에러 응답을 저희가 원하는 예외로 던져주고,** 이 예외를 **서비스에서 핸들링 해주는 것**이 더 좋을 것 같네요~!

#### `ResponseErrorHandler`
> 토스 API 응답에 따른 예외 처리 담당

https://github.com/woowacourse/spring-roomescape-payment/blob/37743b256079a1bf3e614784886cba5ad55ee40e/src/main/java/roomescape/payment/infrastructure/TossPaymentErrorHandler.java#L31-L46

#### `CreateReservationWithPaymentService`
> 예외를 핸들링하고, 이 과정에서 필요한 DB 작업 수행 (롤백 등)

https://github.com/woowacourse/spring-roomescape-payment/blob/37743b256079a1bf3e614784886cba5ad55ee40e/src/main/java/roomescape/reservation/service/CreateReservationWithPaymentService.java#L33-L47

---

이렇게 바꾸니까 객체의 책임도 더 명확하게 분리되는 것 같습니다!
> - 토스 페이먼츠 API 응답 예외는 모두 `ResponseErrorHandler`가 처리
> - `TossPaymentClient`는 외부 API를 호출하고, 의존성 주입받은 `ResponseErrorHandler` 구현체를 매핑

[feat: 토스 결제 승인 API 에러 핸들링 방식 변경 - 검증 메서드 -> onStatus 에러 핸들러로 변경](https://github.com/woowacourse/spring-roomescape-payment/pull/211/commits/02b366c9e2a3c09c64c7babde18a91fb4068bfe2)

✚ 리팩터링을 통해 테스트 코드에서 `@RestClientTest` 어노테이션도 새롭게 학습하고 적용해볼 수 있었어요!

Gomding reviewed (changes requested):
안녕하세요 밍곰!
피드백 적절하게 반영해주셨네요 💯 💯 
추가로 코멘트 남겼으니 확인 부탁드려요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요~!

minSsan committed:
feat(TossPaymentClient): 멱등키 헤더로 인해 발생하는 오류 처리

- 관련 문서 링크: https://docs.tosspayments.com/reference/using-api/authorization#%EC%97%90%EB%9F%AC-%EC%B2%98%EB%A6%AC%ED%95%98%EA%B8%B0

minSsan committed:
feat(TossPaymentClient): confirm 요청 retry 조건 명시 - 응답이 5XX 에러인 경우만 재시도

minSsan committed:
feat: timeout 설정을 yml 파일에서 관리하도록 변경

- timeout은 서버 환경, 물리적인 환경에 따라서 달라질 수 있다.
- 즉, 분산 환경의 경우 서버마다 설정 값이 달라질 수 있다.

minSsan committed:
feat: toss api connection timeout 변경 - 3초 -> 1초

minSsan committed:
chore(PaymentStatus): 불필요한 enum 생성자 괄호 삭제

Co-authored-by: 박민영 <57378410+Gomding@users.noreply.github.com>

minSsan committed:
refactor: TossPaymentClient 를 관리하는 infrastructure 위치 변경 - payment 도메인 패키지로 이동

minSsan committed:
refactor(AuthHeaderEncoder): 패키지 위치 이동

minSsan committed:
refactor: RestClientConfig -> PaymentRestClientConfig 이름 변경

minSsan committed:
feat: 토스 결제 승인 API 에러 핸들링 방식 변경 - 검증 메서드 -> onStatus 에러 핸들러로 변경

minSsan committed:
test(TossPaymentClient): 토스 페이먼츠 API 클라이언트 객체 테스트 작성

minSsan reviewed (commented):
안녕하세요 찰리! 우선 좋은 피드백 남겨주셔서 감사합니다 🙇🏻‍♀️

리뷰 반영 과정에서 기존 테스트 몇 개가 터져서, 관련 개념을 학습하고 적용하느라 리뷰 요청이 조금 늦어졌습니다 😥 죄송합니다 ㅜㅜ

---

몇몇 코멘트는 `single comment`로 바로 제출해버려서, 혹시나 누락이 될까 싶어서 링크로 첨부하겠습니다..!

https://github.com/woowacourse/spring-roomescape-payment/pull/211#discussion_r2123675262

Gomding commented on src/main/java/roomescape/config/RestClientConfig.java:
```diff
@@ -16,10 +16,17 @@ public class RestClientConfig {
         SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
-        requestFactory.setReadTimeout(15_000);
+        requestFactory.setConnectTimeout(connectionTimeout);
+        requestFactory.setReadTimeout(readTimeout);
```
설정으로 만들어서 세팅해주셨군요 👍 👍

Gomding commented on src/main/resources/application.yml:
```diff
@@ -8,7 +8,7 @@ api:
     timeout:
       read: 15000
-      connection: 3000
+      connection: 1000
```
timeout 의 경우 밍곰 서버의 성능도 고려를 해봐야해요 ㅎㅎ
만약 외부 API 응답이 늦어 스레드를 장시간 사용하게되면 병목이 생기고
인기 방탈출 테마 예약에 사용자가 몰려서 요청이 계속 쌓이면 스레드 풀이 고갈될 수 있겠죠!

그래서 비즈니스 요구사항과 시스템 자원을 고려해서 설정해볼 수 있어요 :)

현재는 변경할 필요는 없습니다 ㅎㅎ

Gomding commented on src/main/java/roomescape/payment/infrastructure/TossPaymentClient.java:
```diff
@@ -38,38 +29,7 @@ public ResponseEntity<ConfirmPaymentResponse> postConfirmPayment(ConfirmPaymentR
                 .header("Idempotency-Key", idempotencyKey.toString())
                 .body(paymentRequest)
                 .retrieve()
+                .onStatus(tossPaymentErrorHandler)
```
RestClient builder 에서 `.defaultStatusHandler()` 로도 설정이 가능해요!
하지만 범용적으로 설정하는 것이기 때문에 장단점이 있죠 ㅎㅎ
밍곰은 어떤 의도로 `onStatus` 에 에러 핸들러를 설정해주셨을까요?

Gomding reviewed (approved):
안녕하세요 밍곰!
치열한 고민과 함께 잘 학습하고 계시네요 👍 
충분히 구현해주셔서 이번 단계는 이만 머지하겠습니다~!!
추가로 궁금한 점 있으면 언제든 DM 남겨주세요 😄
